### PR TITLE
Creates endpoint for fetching student merges

### PR DESCRIPTION
**Pull request by eckermania**

### CODE REVIEW

File: api/src/main/java/ca/bc/gov/educ/eas/api/endpoint/v1/PENRecordEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import io.swagger.v3.oas.annotations.responses.ApiResponses;
+import io.swagger.v3.oas.annotations.tags.Tag;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+
+import java.util.List;
+
+@RequestMapping(URL.BASE_URL + "/pen-records/createDateInRange")
```
Comment (sumathi-thirumani) [inline api/src/main/java/ca/bc/gov/educ/eas/api/endpoint/v1/PENRecordEndpoint.java]: The path identifier createDateInRange could be moved to specific endpoint. Better to have in lowercase.

File: api/src/main/java/ca/bc/gov/educ/eas/api/rest/RestUtils.java [hunk 2]
```diff
@@ -324,4 +327,26 @@ public List<School> getAllSchoolList(UUID correlationID) {
...
   }
+
+  @Retryable(retryFor = {Exception.class}, noRetryFor = {EasAPIRuntimeException.class}, backoff = @Backoff(multiplier = 2, delay = 2000))
+  public List<StudentMerge> getMergedStudentsForDateRange(UUID correlationID, String createDateStart, String createDateEnd) {
+    try {
+      final TypeReference<Event> refEventResponse = new TypeReference<>() {};
+      final TypeReference<List<StudentMerge>> refMergedStudentResponse = new TypeReference<>() {};
+      Object event = Event.builder().sagaId(correlationID).eventType(EventType.GET_MERGED_STUDENT_IDS).eventPayload(CREATE_DATE_START.concat("=").concat(createDateStart).concat("&").concat(CREATE_DATE_END).concat("=").concat(createDateEnd)).build();
+      val responseMessage = this.messagePublisher.requestMessage(TopicsEnum.PEN_SERVICES_API_TOPIC.toString(), JsonUtil.getJsonBytesFromObject(event)).completeOnTimeout(null, 120, TimeUnit.SECONDS).get();
+      if (responseMessage == null) {
```
Comment (sumathi-thirumani) [inline api/src/main/java/ca/bc/gov/educ/eas/api/rest/RestUtils.java:337]: What will happen to the scenario of no merges happened between the period. Will this condition hold good ?

File: api/src/main/java/ca/bc/gov/educ/eas/api/service/v1/PENRecordService.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    private final RestUtils restUtils;
+
+    @Autowired
+    public PENRecordService(RestUtils restUtils){
+        this.restUtils = restUtils;
+    }
+
+    public List<StudentMerge> getMergedStudentsForDateRange(String createDateStart, String createDateEnd){
+        return restUtils.getMergedStudentsForDateRange(UUID.randomUUID(), createDateStart, createDateEnd);
```
Comment (sumathi-thirumani) [inline api/src/main/java/ca/bc/gov/educ/eas/api/service/v1/PENRecordService.java]: Do we need this randomID ?

Comment (eckermania) [reply api/src/main/java/ca/bc/gov/educ/eas/api/service/v1/PENRecordService.java]: as far as I know, we do - it is only a transaction id that will be shared between the student api and assessments. basically grabbed this directly from how we are retrieving merges in sdc

File: api/src/main/java/ca/bc/gov/educ/eas/api/controller/v1/PENRecordController.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import ca.bc.gov.educ.eas.api.service.v1.PENRecordService;
+import ca.bc.gov.educ.eas.api.struct.v1.StudentMerge;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@Slf4j
+@RestController
+public class PENRecordController implements PENRecordEndpoint {
```
Comment (arcshiftsolutions) [inline api/src/main/java/ca/bc/gov/educ/eas/api/controller/v1/PENRecordController.java]: Please rename this class & service to remove references to PEN. We're retrieving student merges.

File: api/src/main/java/ca/bc/gov/educ/eas/api/endpoint/v1/PENRecordEndpoint.java (hunk 1 above)
Comment (arcshiftsolutions) [reply api/src/main/java/ca/bc/gov/educ/eas/api/endpoint/v1/PENRecordEndpoint.java]: Please modify this to:
`@RequestMapping(URL.BASE_URL + "/student-merges")`

File: api/src/main/java/ca/bc/gov/educ/eas/api/rest/RestUtils.java [hunk 5]
```diff
@@ -324,4 +327,26 @@ public List<School> getAllSchoolList(UUID correlationID) {
       throw new EasAPIRuntimeException(NATS_TIMEOUT + correlationID + ex.getMessage());
     }
   }
+
+  @Retryable(retryFor = {Exception.class}, noRetryFor = {EasAPIRuntimeException.class}, backoff = @Backoff(multiplier = 2, delay = 2000))
```
Comment (arcshiftsolutions) [inline api/src/main/java/ca/bc/gov/educ/eas/api/rest/RestUtils.java]: Please remove this line.

File: api/src/main/java/ca/bc/gov/educ/eas/api/rest/RestUtils.java (hunk 2 above)
Comment (arcshiftsolutions) [reply api/src/main/java/ca/bc/gov/educ/eas/api/rest/RestUtils.java:337]: Empty list is returned from the PEN Services API - should be fine.

File: api/src/main/java/ca/bc/gov/educ/eas/api/service/v1/PENRecordService.java (hunk 3 above)
Comment (arcshiftsolutions) [reply api/src/main/java/ca/bc/gov/educ/eas/api/service/v1/PENRecordService.java]: @eckermania - It's only useful if we log in on this side.

File: api/src/main/java/ca/bc/gov/educ/eas/api/struct/v1/StudentMerge.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+public class StudentMerge extends BaseRequest implements Serializable {
+    /**
+     * The constant serialVersionUID.
+     */
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * The Student merge id.
+     */
+    String studentMergeID;
```
Comment (arcshiftsolutions) [inline api/src/main/java/ca/bc/gov/educ/eas/api/struct/v1/StudentMerge.java:32]: This is also not nullable.

**Review by sumathi-thirumani — APPROVED** [review]

**Pull request merged**